sql/wsrep_mysqld.cc

```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+        }
+    }
+    rc = -1;
+  } else {
+    thd->wsrep_nbo.keys = key_array;
+
+    WSREP_DEBUG(
+        "Query (%s) with write-set (%lld) and exec_mode: %s"
+        " replicated in TO Isolation mode",
+        WSREP_QUERY(thd), (long long)wsrep_thd_trx_seqno(thd),
+        wsrep_thd_client_mode_str(thd));
+
+    snprintf(thd->wsrep_info, sizeof(thd->wsrep_info),
+             "wsrep: TO isolation initiated for write set (%lld)",
+             (long long)wsrep_thd_trx_seqno(thd));
+    WSREP_DEBUG("%s", thd->wsrep_info);
+    thd_proc_info(thd, thd->wsrep_info);
+
+    /* DDL transaction are now atomic so append wsrep xid.
+    This will ensure transactions are logged with the given xid. */
+    bool atomic_ddl = is_atomic_ddl(thd, true);
+    if (atomic_ddl) {
+      wsrep_xid_init(thd->get_transaction()->xid_state()->get_xid(),
+                     thd->wsrep_cs().toi_meta().gtid());
+    }
+
+    ++wsrep_to_isolation;
+    rc = 0;
+  }
+
+  if (buf) my_free(buf);
+
+  thd->wsrep_gtid_event_buf_len = 0;
+  thd->wsrep_gtid_event_buf = NULL;
+
+  if (rc) wsrep_NBO_begin_failed(thd, NULL);
+
+  return rc;
+}
+
+void wsrep_nbo_isolation_begin_end(THD *thd) {
+  wsrep::mutable_buffer err;
+  WSREP_DEBUG("Ending NBO phase one");
+  int ret = thd->wsrep_cs().end_nbo_phase_one(err);
+  WSREP_DEBUG("NBO phase one ended");
+  if (ret) {
+    assert(thd->wsrep_cs().current_error());
+    WSREP_DEBUG("end_nbo_phase_one() failed for %u: %s, seqno: %lld",
+                thd->thread_id(), WSREP_QUERY(thd),
+                (long long)wsrep_thd_trx_seqno(thd));
+  }
+}
+
+int wsrep_nbo_isolation_end_begin(THD *thd) {
+  WSREP_DEBUG("NBO END maybe start");
+  wsrep::client_state &client_state(thd->wsrep_cs());
+  if (wsrep_thd_is_in_nbo(thd)) {
+    WSREP_DEBUG("NBO END start: %lld: %s",
+                client_state.toi_meta().seqno().get(), WSREP_QUERY(thd));
+    WSREP_DEBUG("NBO END start: %lld: %s",
+                client_state.transaction().ws_meta().seqno().get(),
+                WSREP_QUERY(thd));
+    // assert(strstr(WSREP_QUERY(thd), "f1") == 0);
+    int ret = client_state.begin_nbo_phase_two(thd->wsrep_nbo.keys);
+
+    if (ret) {
+      assert(client_state.current_error());
+      // HERE's the cnonconflicting issue
+      WSREP_DEBUG("begin_nbo_phase_two() failed for %u: %s, seqno: %lld",
+                  thd->thread_id(), WSREP_QUERY(thd),
+                  (long long)wsrep_thd_trx_seqno(thd));
+
+      /* jump to error handler in mysql_execute_command() */
+      switch (client_state.current_error()) {
+        default:
+          WSREP_WARN(
+              "NBO isolation failed for: %d, schema: %s, sql: %s. "
+              "Check wsrep connection state and retry the query.",
+              ret, (thd->db().str ? thd->db().str : "(null)"),
+              WSREP_QUERY(thd));
+          if (!thd->is_error()) {
+            WSREP_WARN("Adjusting error to deadlock");
```

Comment:
I don't think this message needs to be logged as a warning. Can we change it to
```suggestion
            WSREP_DEBUG("Adjusting error to deadlock");
```

Reply:
fixed, and removed the commented block below.